storage: extract error from assertReplicaState

tbg opened the pull request:
The 1.0.1 ship may already have sailed, but if it has not,
then let us consider whether we want to get this in to get
more information on #16004 in the wild.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16013)
<!-- Reviewable:end -->

dt commented on pkg/storage/replica.go:
```diff
@@ -1348,7 +1348,13 @@ func (r *Replica) assertStateLocked(ctx context.Context, reader engine.Reader) {
+		// TODO(dt): expose properly once #15892 is addressed.
+		log.Fatal(
+			ctx, log.Safe{
+				V: fmt.Sprintf("on-disk and in-memory state diverged:\n%s", pretty.Diff(diskState, r.mu.state)),
```
eh, this would potentially include ReplicaState.Desc which has keys in it, no?

tbg committed:
storage: extract error from assertReplicaState

The 1.0.1 ship may already have sailed, but if it has not,
then let us consider whether we want to get this in to get
more information on #16004 in the wild.

tschottdorf commented:
That's a good point, changed. Nothing else in ReplicaState seems sensitive (mostly lease and numbers). What would this look like after the RFC?

dt commented:
Well, the RFC is initially pretty conservative, so pretty much like this initially -- we should default to assuming `pretty.Diff` contains sensitive data unless we're certain otherwise. After the initial RFC implementation, which should get most of the low-hanging fruit in terms of common string errors, we can look at tagging struct fields and using our own formatter so you could potentially get something other than total redaction of a %v with a non-safe struct arg, 

---

Reviewed 1 of 2 files at r2.
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/log/crash_reporting.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16013#-KkQp6RzjZ6O5KM15Yqt:-KkQp6RzjZ6O5KM15Yqu:bkwb7g3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6abb0e730e2ee4fbef5a917199582b355ecc5526/pkg/util/log/crash_reporting.go#L92)):*
> ```Go
> 	file, line, _ := caller.Lookup(depth + 3)
> 
> 	reportable := fmt.Sprintf("%s %s:%d", format(r), filepath.Base(file), line)
> ```

I'm not sure the file:line is strictly needed when we're printing the actual error - the report includes the full stack traces anyway -- it was just being included with the `%T` so we could tell two `errors.fundamental` or `errors.wrappedError` apart without needing to look at their stacks.

That said, if you just want to get this merged for 1.0.1, I'm happy to fix along with the rest of the full implementation of the RFC.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16013#-:-KkQpw_C0pHNi5z275Y3:b-1z9kg1)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1356 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16013#-KkQrAEjWuuFgqiLHwNR:-KkQrAEjWuuFgqiLHwNS:bhxmgwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6abb0e730e2ee4fbef5a917199582b355ecc5526/pkg/storage/replica.go#L1356)):*
> ```Go
> 		log.Errorf(ctx, "on-disk and in-memory state diverged:\n%s", pretty.Diff(diskState, r.mu.state))
> 		r.mu.state.Desc, diskState.Desc = nil, nil
> 		log.Fatal(
> ```

alternative proposal, since now you're double-logging it locally, that also avoids front-running any details of #15892 :
if you really just want this particular report, you could just export `sendCrashReport` and then call it directly with what you want reported to sentry (it assumes its args are already redacted and ready to be reported).

We'd get both reports in sentry and just ignore the less detailed one, but the user's logs wouldn't get spammed as much.

I don't have particularly strong opinions though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16013)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16013#-:-KkRPGLHYwRDaYFPd2sz:bnfp4nl)*
<!-- Sent from Reviewable.io -->
